File: library/src/main/java/com/xiaoyv/javaengine/JavaEngine.java
package com.xiaoyv.javaengine;

import android.app.Application;
import android.util.Log;

import com.xiaoyv.javaengine.compile.ClassCompiler;
import com.xiaoyv.javaengine.compile.DexCompiler;
import com.xiaoyv.javaengine.executor.DexExecutor;
import com.xiaoyv.javaengine.utils.FileUtils;
import com.xiaoyv.javaengine.utils.ResourceUtils;
import com.xiaoyv.javaengine.utils.Utils;

/**
 * @author 王怀玉 编译器管理类
 * @since 2020/5/10
 */
public class JavaEngine {

    /**
     * 初始化 Java编译器引擎
     */
    public static void init(Application application) {
        Utils.init(application);
        boolean b = checkRtJar();
        Log.i("checkRtJar", "Java Runtime 编译文件----复制：" + b);
    }

    /**
     * 检查 rt.jar 是否存在
     */
    public static boolean checkRtJar() {
        String rtPath = JavaEngineSetting.getRtPath();
        // 不存在时从 Assets 复制
        boolean fileExists = FileUtils.isFileExists(rtPath);
        if (!fileExists) {
            return ResourceUtils.copyFileFromAssets("rt.jar", rtPath);
        }
        return true;
    }


    /**
     * Class 编译器
     */
    private volatile static ClassCompiler classCompiler;

    public static ClassCompiler getClassCompiler() {
        if (classCompiler == null) {
            synchronized (ClassCompiler.class) {
                if (classCompiler == null) {
                    classCompiler = new ClassCompiler();
                }
            }
        }
        return classCompiler;
    }


    /**
     * Dex 编译器
     */
    private volatile static DexCompiler dexCompiler;

    public static DexCompiler getDexCompiler() {
        if (dexCompiler == null) {
            synchronized (DexCompiler.class) {
                if (dexCompiler == null) {
                    dexCompiler = new DexCompiler();
                }
            }
        }
        return dexCompiler;
    }

    /**
     * Dex 执行器
     */
    private volatile static DexExecutor dexExecutor;

    public static DexExecutor getDexExecutor() {
        if (dexExecutor == null) {
            synchronized (DexExecutor.class) {
                if (dexExecutor == null) {
                    dexExecutor = new DexExecutor();
                }
            }
        }
        return dexExecutor;
    }
}

File: compiler-d8/src/main/java/com/xiaoyv/java/compiler/JavaEngine.kt
package com.xiaoyv.java.compiler

import android.app.Application
import android.util.Log
import com.xiaoyv.java.compiler.dex.JavaDexCompiler
import com.xiaoyv.java.compiler.java.JavaClassCompiler
import com.xiaoyv.java.compiler.utils.GlobalUtils
import com.xiaoyv.java.compiler.utils.ResourceUtils
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import java.io.File

/**
 * JavaEngine
 *
 * @author why
 * @since 2022/3/7
 */
object JavaEngine {
    private const val TAG = "JavaEngine"

    val CompileExceptionHandler = CoroutineExceptionHandler { coroutineContext, throwable ->
        logError(throwable)
    }

    /**
     * 编译器设置
     */
    @JvmStatic
    val compilerSetting: JavaEngineSetting by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        JavaEngineSetting()
    }

    /**
     * Class 编译器
     */
    @JvmStatic
    val classCompiler: JavaClassCompiler by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        JavaClassCompiler()
    }

    /**
     * Dex 编译器
     */
    @JvmStatic
    val dexCompiler: JavaDexCompiler by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        JavaDexCompiler()
    }

    /**
     * 初始化
     */
    @JvmStatic
    fun init(application: Application) {
        GlobalUtils.init(application)

        CoroutineScope(Dispatchers.IO).launch {
            installRtJar().collect {
                Log.i(TAG, "Rt.jar 安装结果：$it")
            }
        }
    }

    /**
     * 安装 Rt.jar
     */
    private suspend fun installRtJar(): Flow<Boolean> = flow {
        val rtJar = File(compilerSetting.rtPath)

        // 不存在时从 Assets 复制
        val fileExists = rtJar.exists() && rtJar.length() != 0L
        if (fileExists) {
            emit(true)
        } else {
            emit(ResourceUtils.copyFileFromAssets("rt.jar", rtJar.absolutePath))
        }
    }

    internal fun logInfo(msg: String) {
        Log.i(TAG, msg)
    }

    internal fun logError(msg: String) {
        Log.e(TAG, msg)
    }

    internal fun logError(e: Throwable) {
        Log.e(TAG, e.toString(), e)
    }
}